File: src/01-fundamentals/01-exercises/02-immutability/SmartImmutability.scala
/**
 * =============================================================================
 * 🧠 INMUTABILIDAD INTELIGENTE: EL SECRETO DE LA EFICIENCIA EN FP
 * =============================================================================
 *
 * Referencia: "Functional Programming in Scala" - Capítulo 3 (Data Structures)
 *
 * En Scala, la inmutabilidad está integrada en el lenguaje.
 * Este archivo demuestra cómo Scala implementa la compartición estructural
 * de forma nativa y segura.
 *
 * =============================================================================
 */

object SmartImmutability extends App {

  // ===========================================================================
  // CONCEPTO 1: Lista Inmutable como Sealed Trait
  // ===========================================================================

  /**
   * 🔒 DEFINICIÓN: Lista Enlazada Inmutable (como en FP in Scala)
   *
   * Una lista puede ser:
   * - Nil: lista vacía
   * - Cons(head, tail): un elemento seguido del resto de la lista
   *
   * 'sealed' garantiza que no hay más subtipos fuera de este archivo,
   * permitiendo pattern matching exhaustivo.
   */
  sealed trait ImmutableList[+A]  // +A = covariante (explicado más adelante)

  case object Nil extends ImmutableList[Nothing]  // Nothing = subtipo de todos

  case class Cons[+A](head: A, tail: ImmutableList[A]) extends ImmutableList[A]

  // Objeto companion con operaciones de lista
  object ImmutableList {

    // Constructor desde varargs: ImmutableList(1, 2, 3)
    def apply[A](as: A*): ImmutableList[A] =
      if (as.isEmpty) Nil
      else Cons(as.head, apply(as.tail: _*))

    // Convertir a Scala List (para imprimir fácilmente)
    def toScalaList[A](list: ImmutableList[A]): List[A] = list match {
      case Nil => List.empty
      case Cons(h, t) => h :: toScalaList(t)
    }
  }

  import ImmutableList._

  // ===========================================================================
  // CONCEPTO 2: COMPARTICIÓN ESTRUCTURAL
  // ===========================================================================

  /**
   * 🔗 PREPEND: Agregar al inicio - O(1)
   *
   * Solo creamos UN nodo nuevo que apunta a la lista existente.
   * No copiamos nada de la estructura original.
   */
  def prepend[A](element: A, list: ImmutableList[A]): ImmutableList[A] =
    Cons(element, list)  // O(1) - crea solo un nodo

  /**
   * 📐 TAIL: Obtener el resto - O(1)
   *
   * Simplemente retornamos la referencia existente.
   * El tail ya existe como parte de la estructura.
   */
  def tail[A](list: ImmutableList[A]): ImmutableList[A] = list match {
    case Nil => throw new NoSuchElementException("tail de lista vacía")
    case Cons(_, t) => t  // O(1) - solo retorna referencia existente
  }

  /**
   * 🔍 HEAD: Obtener el primer elemento - O(1)
   */
  def head[A](list: ImmutableList[A]): A = list match {
    case Nil => throw new NoSuchElementException("head de lista vacía")
    case Cons(h, _) => h
  }

  // ===========================================================================
  // CONCEPTO 3: EFICIENCIA EN LA PRÁCTICA
  // ===========================================================================

  /**
   * ⚡ DEMOSTRACIÓN: Prepend es O(1)
   */
  def demoPrependRapido(): Unit = {
    println("⚡ OPERACIÓN RÁPIDA: Agregar al inicio (Prepend)")
    println("─" * 50)

    // Lista original: [X, Y, Z]
    val original = ImmutableList("X", "Y", "Z")
    println(s"Lista original: ${toScalaList(original)}")

    // Agregar "New" al inicio - O(1)
    val conNew = prepend("New", original)
    println(s"Con 'New' al inicio: ${toScalaList(conNew)}")

    // La lista original sigue INTACTA
    println(s"Lista original (sin cambios): ${toScalaList(original)}")

    println("\n📊 Análisis:")
    println("   - Creamos solo 1 nodo nuevo (Cons)")
    println("   - Reutilizamos toda la lista original como tail")
    println("   - Complejidad: O(1) tiempo constante")
    println()
  }

  /**
   * 🐢 INIT: Quitar el último elemento - O(n)
   *
   * Necesitamos reconstruir toda la lista excepto el último elemento,
   * porque no podemos modificar los nodos existentes.
   */
  def init[A](list: ImmutableList[A]): ImmutableList[A] = list match {
    case Nil => throw new NoSuchElementException("init de lista vacía")
    case Cons(_, Nil) => Nil  // Solo un elemento -> retornar lista vacía
    case Cons(h, t) => Cons(h, init(t))  // Reconstruir recursivamente O(n)
  }

  def demoInitLento(): Unit = {
    println("🐢 OPERACIÓN LENTA: Quitar del final (Init)")
    println("─" * 50)

    val original = ImmutableList(1, 2, 3, 4)
    println(s"Lista original: ${toScalaList(original)}")

    val sinUltimo = init(original)
    println(s"Sin el último: ${toScalaList(sinUltimo)}")

    println("\n📊 Análisis:")
    println("   - Copiamos 3 nodos (todos excepto el eliminado)")
    println("   - Cada llamada recursiva crea un nuevo Cons")
    println("   - Complejidad: O(n) donde n = longitud de la lista")
    println()
  }

  /**
   * 📐 DEMOSTRACIÓN: Tail es O(1)
   */
  def demoTail(): Unit = {
    println("📐 EJEMPLO VISUAL: 'tail' de una Lista")
    println("─" * 50)

    val original = ImmutableList("Head", "T", "A", "I", "L")
    println(s"Lista original: ${toScalaList(original)}")

    val tailDeLista = tail(original)
    println(s"Tail de la lista: ${toScalaList(tailDeLista)}")

    println(s"Original (sin cambios): ${toScalaList(original)}")

    println("\n💡 Clave:")
    println("   - tail() NO copia nada")
    println("   - Solo pattern-match y retorna la referencia existente")
    println("   - ¡Esta parte de la lista ya existía!")
    println()
  }

  // ===========================================================================
  // CONCEPTO 4: ADIÓS A LAS COPIAS DEFENSIVAS
  // ===========================================================================

  /**
   * 🛡️ EN SCALA, la inmutabilidad elimina la necesidad de copias defensivas
   *
   * Con case class + val + colecciones inmutables, es IMPOSIBLE modificar
   * los datos compartidos. El compilador lo garantiza.
   */

  // Tipos inmutables (case class = inmutable por defecto)
  case class Usuario(id: String, nombre: String)

  // Repositorio inmutable
  case class RepositorioInmutable(usuarios: ImmutableList[Usuario] = Nil) {

    // Retorna un NUEVO repositorio con el usuario agregado
    def agregarUsuario(id: String, nombre: String): RepositorioInmutable =
      RepositorioInmutable(prepend(Usuario(id, nombre), usuarios))

    // ✅ Podemos retornar la referencia sin miedo - es inmutable
    def getUsuarios: ImmutableList[Usuario] = usuarios
  }

  def demoCopiasDefensivas(): Unit = {
    println("🛡️ ADIÓS A LAS COPIAS DEFENSIVAS")
    println("─" * 50)

    println("\n✅ EN SCALA (inmutable por diseño):")

    val repo1 = RepositorioInmutable()
    val repo2 = repo1.agregarUsuario("1", "Ana")
    val repo3 = repo2.agregarUsuario("2", "Bob")

    println(s"Repo1 (vacío): ${toScalaList(repo1.getUsuarios)}")
    println(s"Repo2 (con Ana): ${toScalaList(repo2.getUsuarios)}")
    println(s"Repo3 (con Ana y Bob): ${toScalaList(repo3.getUsuarios)}")

    // Intentar modificar es un ERROR DE COMPILACIÓN:
    // repo3.usuarios = Nil                    // ❌ val no reassignable
    // repo3.getUsuarios.head.nombre = "X"     // ❌ case class inmutable

    println("\n💡 En Scala:")
    println("   - case class + val = inmutable por default")
    println("   - No existe 'getUsuariosSinProteccion' vs 'getUsuariosConCopia'")
    println("   - Todo es seguro automáticamente")
    println()
  }

  // ===========================================================================
  // CONCEPTO 5: COMPARATIVA TypeScript vs Scala
  // ===========================================================================

  def comparativaLenguajes(): Unit = {
    println("🔄 COMPARATIVA: TypeScript vs Scala")
    println("═" * 65)
    println("│ Aspecto                │ TypeScript          │ Scala             │")
    println("├────────────────────────┼─────────────────────┼───────────────────┤")
    println("│ Inmutabilidad default  │ No (requiere effort)│ Sí (val,case class│")
    println("│ readonly               │ Solo compile-time   │ Runtime enforced  │")
    println("│ Pattern matching       │ Limitado            │ Nativo y poderoso │")
    println("│ Covarianza             │ Inferida            │ Explícita (+A)    │")
    println("│ sealed types           │ No nativo           │ Garantiza exhaust.│")
    println("│ Colecciones inmutables │ Requiere disciplina │ Default del stdlib│")
    println("═" * 65)
    println()
  }

  // ===========================================================================
  // TABLA DE COMPLEJIDAD
  // ===========================================================================

  def tablaComplejidad(): Unit = {
    println("📊 TABLA DE COMPLEJIDAD - Lista Inmutable")
    println("═" * 65)
    println("│ Operación              │ Lista Inmutable │ Array Mutable │")
    println("├────────────────────────┼─────────────────┼───────────────┤")
    println("│ Agregar al INICIO (::) │      O(1)       │     O(n)      │")
    println("│ Agregar al FINAL       │      O(n)       │     O(1)*     │")
    println("│ tail (quitar inicio)   │      O(1)       │     O(n)      │")
    println("│ init (quitar final)    │      O(n)       │     O(1)      │")
    println("│ Acceso por índice      │      O(n)       │     O(1)      │")
    println("│ Compartir datos        │      O(1)       │     O(n)**    │")
    println("═" * 65)
    println("* Amortizado")
    println("** Requiere copia defensiva para seguridad")
    println()
  }

  // ===========================================================================
  // DEMOSTRACIÓN PRINCIPAL
  // ===========================================================================

  println("═" * 70)
  println("🧠 INMUTABILIDAD INTELIGENTE EN SCALA")
  println("═" * 70)
  println()

  demoPrependRapido()
  demoInitLento()
  demoTail()
  tablaComplejidad()
  comparativaLenguajes()
  demoCopiasDefensivas()

  println("═" * 70)
  println("📚 RESUMEN DE CONCEPTOS DEL GRÁFICO:")
  println("═" * 70)
  println("""
1. ESTRUCTURAS INMUTABLES EN SCALA
   → case class + val = inmutable por defecto
   → sealed trait permite pattern matching exhaustivo
   → El compilador GARANTIZA inmutabilidad

2. COMPARTICIÓN ESTRUCTURAL
   → Cons(head, tail) solo guarda referencias
   → prepend (::) es O(1) y reutiliza toda la lista
   → Scala List implementa esto nativamente

3. EFICIENCIA PRÁCTICA
   → head/tail: O(1) - acceso inmediato
   → last/init: O(n) - requiere recorrer/copiar
   → Diseña algoritmos que trabajen del frente hacia atrás

4. SCALA vs TYPESCRIPT
   → Scala: inmutabilidad es el DEFAULT del lenguaje
   → TypeScript: requiere disciplina (readonly, const, etc.)
   → Scala detecta errores en COMPILE-TIME
  """)
  println("═" * 70)

  // ===========================================================================
  // RETO DE REFACTORIZACIÓN
  // ===========================================================================

  println("\n🎯 RETO DE REFACTORIZACIÓN:")
  println("─" * 70)
  println("""
Tienes este código IMPERATIVO. Piénsalo de forma funcional:

❌ CÓDIGO MALO (mutable con var):

var historial = List.empty[String]

def registrar(cambio: String): Unit = {
    historial = historial :+ cambio  // Lento O(n)
}

def deshacer(): Unit = {
    historial = historial.init  // Lento O(n)
}

💭 PREGUNTAS:
   1. ¿Por qué :+ (append) es O(n) pero :: (prepend) es O(1)?
   2. ¿Cómo reescribirías esto para que ambas operaciones sean O(1)?
   3. ¿Necesitas var o puedes hacerlo completamente inmutable?

✅ PISTA: Usa prepend para "registrar" y tail para "deshacer".
   El historial se lee en orden inverso (el más reciente primero).
  """)
  println("─" * 70)
}

File: src/01-fundamentals/01-exercises/02-immutability/smart-immutability.ts
/**
 * =============================================================================
 * 🧠 INMUTABILIDAD INTELIGENTE: EL SECRETO DE LA EFICIENCIA EN FP
 * =============================================================================
 *
 * Referencia: "Functional Programming in Scala" - Capítulo 3 (Data Structures)
 *
 * En la programación funcional, las estructuras de datos son INMUTABLES.
 * Pero, ¿no es eso terriblemente ineficiente? ¿Copiar todo cada vez?
 *
 * ¡NO! El secreto es la "COMPARTICIÓN ESTRUCTURAL" (structural sharing).
 * Las operaciones crean NUEVAS estructuras que REUTILIZAN la mayor parte
 * de los datos originales, en lugar de copiarlos.
 *
 * =============================================================================
 */

// =============================================================================
// CONCEPTO 1: ¿Qué es una Estructura de Datos Inmutable?
// =============================================================================

/**
 * 🔒 ANALOGÍA: Piensa en un DOCUMENTO SELLADO
 *
 * Imagina un documento legal que, una vez firmado y sellado, no puede ser
 * modificado. Si necesitas hacer un cambio, debes crear un NUEVO documento.
 *
 * Pero aquí está el truco: puedes REFERENCIAR partes del documento original
 * sin copiarlo todo. Por ejemplo, "Ver anexo del documento original".
 *
 * Esto es exactamente lo que hacen las estructuras de datos inmutables.
 */

// Una Lista Enlazada Inmutable simple (como la de FP in Scala)
// Cada nodo solo tiene un valor y una referencia al resto de la lista
type ImmutableList<A> = Nil | Cons<A>;

// Nil representa la lista vacía
interface Nil {
  readonly _tag: 'Nil';
}

// Cons representa un nodo con un valor y el resto de la lista
interface Cons<A> {
  readonly _tag: 'Cons';
  readonly head: A; // El valor de este nodo
  readonly tail: ImmutableList<A>; // Referencia al resto (¡no una copia!)
}

// Constructores helper
const nil: Nil = { _tag: 'Nil' };

function cons<A>(head: A, tail: ImmutableList<A>): Cons<A> {
  return { _tag: 'Cons', head, tail };
}

// Verificadores de tipo
function isNil<A>(list: ImmutableList<A>): list is Nil {
  return list._tag === 'Nil';
}

function isCons<A>(list: ImmutableList<A>): list is Cons<A> {
  return list._tag === 'Cons';
}

// =============================================================================
// CONCEPTO 2: COMPARTICIÓN ESTRUCTURAL (Structural Sharing)
// =============================================================================

/**
 * 🔗 PRINCIPIO CLAVE: Reutilización en Lugar de Copia
 *
 * Cuando creamos una nueva lista, NO copiamos los nodos existentes.
 * ¡Solo creamos NUEVAS REFERENCIAS a ellos!
 *
 * Ejemplo:
 *   listaOriginal = [A] -> [B] -> [C] -> Nil
 *
 *   Al agregar "N" al inicio:
 *   nuevaLista = [N] -> [A] -> [B] -> [C] -> Nil
 *                        ↑
 *                        │
 *   listaOriginal ───────┘ (¡sigue apuntando a [A]!)
 *
 *   Solo creamos UN nuevo nodo [N], el resto se REUTILIZA.
 */

// Crear lista desde array (helper)
function fromArray<A>(arr: readonly A[]): ImmutableList<A> {
  // Construimos la lista de derecha a izquierda
  // Empezamos con nil y vamos agregando al frente
  let result: ImmutableList<A> = nil;
  for (let i = arr.length - 1; i >= 0; i--) {
    result = cons(arr[i], result);
  }
  return result;
}

// Convertir lista a array (helper)
function toArray<A>(list: ImmutableList<A>): A[] {
  const result: A[] = [];
  let current = list;
  while (isCons(current)) {
    result.push(current.head);
    current = current.tail;
  }
  return result;
}

// Agregar al inicio: O(1) - ¡Solo creamos un nodo!
function prepend<A>(element: A, list: ImmutableList<A>): ImmutableList<A> {
  // Este es el corazón de la compartición estructural:
  // Creamos UN solo nodo nuevo que APUNTA a toda la lista existente
  // No copiamos nada de la lista original
  return cons(element, list); // O(1) - tiempo constante
}

// =============================================================================
// CONCEPTO 3: EFICIENCIA EN LA PRÁCTICA
// =============================================================================

/**
 * ⚡ RÁPIDO: Agregar al Principio (prepend) - O(1)
 *
 * Solo creamos un nuevo nodo que apunta a la lista existente.
 *
 * [New] -> [X] -> [Y] -> [Z] -> Nil
 *           ↑
 *           └── La lista original se reutiliza completa
 */
function demoPrependRapido(): void {
  console.log('⚡ OPERACIÓN RÁPIDA: Agregar al inicio (Prepend)');
  console.log('─'.repeat(50));

  // Lista original: [X, Y, Z]
  const original = fromArray(['X', 'Y', 'Z']);
  console.log('[original]', original);
  console.log('Lista original:', toArray(original));
  const a_array = toArray(original);
  console.log('[a_array]', a_array);
  console.log('[a_array]', Array.isArray(a_array));
  console.log('[a_array]', typeof a_array);

  // Agregar "New" al inicio - O(1)
  const conNew = prepend('New', original);
  console.log('[conNew]', conNew);
  console.log("Con 'New' al inicio:", toArray(conNew));

  // ¡La lista original sigue INTACTA!
  console.log('Lista original (sin cambios):', toArray(original));

  console.log('\n📊 Análisis:');
  console.log('   - Creamos solo 1 nodo nuevo');
  console.log('   - Reutilizamos 3 nodos existentes');
  console.log('   - Complejidad: O(1) tiempo constante');
  console.log();
}

/**
 * 🐢 LENTO: Quitar del Final (init) - O(n)
 *
 * Necesitamos crear copias de TODOS los nodos excepto el último,
 * porque cada nodo apunta al siguiente y no puede ser modificado.
 *
 * Original: [1] -> [2] -> [3] -> [4] -> Nil
 *
 * Para quitar [4]:
 * Nuevo:    [1'] -> [2'] -> [3'] -> Nil
 *
 * Cada [n'] es un NUEVO nodo (copia del valor, nueva referencia)
 */
function init<A>(list: ImmutableList<A>): ImmutableList<A> {
  if (isNil(list)) {
    throw new Error('init de lista vacía');
  }
  // Si solo queda un elemento, retornamos lista vacía
  if (isNil(list.tail)) {
    return nil;
  }
  // Recursivamente: mantenemos head, aplicamos init al tail
  // Esto crea una COPIA de cada nodo (O(n) total)
  return cons(list.head, init(list.tail));
}

function demoInitLento(): void {
  console.log('🐢 OPERACIÓN LENTA: Quitar del final (Init)');
  console.log('─'.repeat(50));

  // Lista original: [1, 2, 3, 4]
  const original = fromArray([1, 2, 3, 4]);
  console.log('[original]', original);
  // [
  //   '[original]', {
  //     _tag: 'Cons',
  //     head: 1,
  //     tail: {
  //       _tag: 'Cons',
  //       head: 2,
  //       tail: {
  //         _tag: 'Cons',
  //         head: 3,
  //         tail: { _tag: 'Cons', head: 4, tail: {...} }
  //       }
  //     }
  //   }
  // ]
  console.log('Lista original:', toArray(original));

  // Quitar el último elemento - O(n)
  const sinUltimo = init(original);
  console.log('Sin el último:', toArray(sinUltimo));

  console.log('\n📊 Análisis:');
  console.log('   - Copiamos 3 nodos (todos excepto el eliminado)');
  console.log('   - No pudimos reutilizar nada de la estructura');
  console.log('   - Complejidad: O(n) donde n = longitud de la lista');
  console.log();
}

// =============================================================================
// CONCEPTO 4: EJEMPLO VISUAL - 'tail' de una Lista
// =============================================================================

/**
 * 📐 Obtener la "cola" (tail) de una lista
 *
 * tail([Head] -> [A] -> [I] -> [L] -> Nil)
 *      └─────────────────────────────────────────→ retorna referencia directa
 *               [A] -> [I] -> [L] -> Nil
 *
 * NO crea copias - solo retorna la referencia existente.
 * El tail de una lista ya existe como parte de la estructura.
 */
function tail<A>(list: ImmutableList<A>): ImmutableList<A> {
  if (isNil(list)) {
    throw new Error('tail de lista vacía');
  }
  // Simplemente retornamos la referencia al tail existente
  // ¡No copiamos nada! La estructura ya existe.
  return list.tail; // O(1) - tiempo constante
}

function head<A>(list: ImmutableList<A>): A {
  if (isNil(list)) {
    throw new Error('head de lista vacía');
  }
  return list.head;
}

function demoTail(): void {
  console.log("📐 EJEMPLO VISUAL: 'tail' de una Lista");
  console.log('─'.repeat(50));

  // Original: [Head] -> [T] -> [A] -> [I] -> [L] -> Nil
  const original = fromArray(['Head', 'T', 'A', 'I', 'L']);
  console.log('Lista original:', toArray(original));

  // Obtener tail - O(1), solo retorna una referencia existente
  const tailDeLista = tail(original);
  console.log('Tail de la lista:', toArray(tailDeLista));

  // La lista original SIGUE INTACTA
  console.log('Original (sin cambios):', toArray(original));

  console.log('\n💡 Clave:');
  console.log('   - tail() NO copia nada');
  console.log('   - Solo retorna la referencia al resto de la lista');
  console.log('   - ¡Esta parte de la lista ya existía!');
  console.log();
}

// =============================================================================
// CONCEPTO 5: ADIÓS A LAS COPIAS DEFENSIVAS
// =============================================================================

/**
 * 🛡️ COPIAS DEFENSIVAS: El problema mutable
 *
 * En programación imperativa, cuando pasas objetos mutables a otra función,
 * debes hacer "copias defensivas" para evitar que la función los modifique.
 *
 * Con inmutabilidad, esto es INNECESARIO porque nadie puede modificar los datos.
 */

// ❌ ENFOQUE MUTABLE: Necesita copias defensivas
class RepositorioMutable {
  private usuarios: { id: string; nombre: string }[] = [];

  agregarUsuario(id: string, nombre: string): void {
    this.usuarios.push({ id, nombre });
  }

  // ⚠️ ¡PELIGRO! Si retornamos la referencia directa,
  // el código externo puede modificar nuestra lista
  getUsuariosSinProteccion(): { id: string; nombre: string }[] {
    return this.usuarios; // ❌ Expone el estado interno
  }

  // ✅ "Copia defensiva" - crear una nueva copia para proteger
  getUsuariosConCopia(): { id: string; nombre: string }[] {
    // Copiamos TODO para proteger el estado interno
    // Esto es O(n) en tiempo y memoria
    return this.usuarios.map(u => ({ ...u }));
  }
}

const testU = new RepositorioMutable();
testU.agregarUsuario('1', 'Ana');
testU.agregarUsuario('2', 'Bob');
const testUSinProteccion = testU.getUsuariosConCopia();
testUSinProteccion.push({ id: '3', nombre: 'hack' });

// console.log('Usuarios originales:', testU.getUsuariosConCopia());

// ✅ ENFOQUE INMUTABLE: No necesita copias defensivas
type UsuarioInmutable = {
  readonly id: string;
  readonly nombre: string;
};

class RepositorioInmutable {
  private readonly usuarios: ImmutableList<UsuarioInmutable>;

  constructor(usuarios: ImmutableList<UsuarioInmutable> = nil) {
    this.usuarios = usuarios;
  }

  // Retorna un NUEVO repositorio con el usuario agregado
  agregarUsuario(id: string, nombre: string): RepositorioInmutable {
    const nuevoUsuario: UsuarioInmutable = { id, nombre };
    return new RepositorioInmutable(prepend(nuevoUsuario, this.usuarios));
  }

  // ✅ Podemos retornar la referencia directa sin miedo
  // Nadie puede modificar una estructura inmutable
  getUsuarios(): ImmutableList<UsuarioInmutable> {
    return this.usuarios; // ¡Seguro! No se puede modificar
  }
}

const testUI = new RepositorioInmutable();
const l1 = testUI.agregarUsuario('1', 'Ana');
const l2 = testUI.agregarUsuario('2', 'Bob');
const testUInmutable = testUI.getUsuarios();
// console.log('[l2]', toArray(l2.getUsuarios()));
// testUInmutable.push({ id: '3', nombre: 'hack' });

// console.log('Usuarios originales:', testUI.getUsuarios());

function demoCopiasDefensivas(): void {
  console.log('🛡️ ADIÓS A LAS COPIAS DEFENSIVAS');
  console.log('─'.repeat(50));

  console.log('\n❌ PROBLEMA CON MUTABILIDAD:');
  const repoMutable = new RepositorioMutable();
  repoMutable.agregarUsuario('1', 'Ana');
  repoMutable.agregarUsuario('2', 'Bob');

  // Sin copia defensiva, el código externo puede romper el estado interno
  const listaInsegura = repoMutable.getUsuariosSinProteccion();
  console.log('Usuarios originales:', listaInsegura);

  // 😱 ¡Modificamos el estado interno del repositorio!
  listaInsegura.push({ id: 'HACKER', nombre: 'Intruso' });
  listaInsegura[0].nombre = 'ANA MODIFICADA';

  console.log('Después de la manipulación externa:');
  console.log('  Estado interno corrupto:', repoMutable.getUsuariosSinProteccion());

  console.log('\n✅ CON INMUTABILIDAD:');
  let repoInmutable = new RepositorioInmutable();
  repoInmutable = repoInmutable.agregarUsuario('1', 'Ana');
  repoInmutable = repoInmutable.agregarUsuario('2', 'Bob');

  // Podemos compartir la referencia sin miedo
  const listaSegura = repoInmutable.getUsuarios();
  console.log('[listaSegura]', listaSegura);
  console.log('[listaSegura]', listaSegura);
  console.log('Usuarios:', toArray(listaSegura));

  // No hay forma de modificar listaSegura - es inmutable
  
  // listaSegura.head.nombre = "X";  // ❌ Error de TypeScript
  // listaSegura._tag = "Nil";       // ❌ Error de TypeScript

  console.log('Estado interno protegido:', toArray(repoInmutable.getUsuarios()));
  console.log();
}

// =============================================================================
// TABLA COMPARATIVA DE COMPLEJIDAD
// =============================================================================

function tablaComplejidad(): void {
  console.log('📊 TABLA DE COMPLEJIDAD - Lista Inmutable vs Array Mutable');
  console.log('═'.repeat(65));
  console.log('│ Operación              │ Lista Inmutable │ Array Mutable │');
  console.log('├────────────────────────┼─────────────────┼───────────────┤');
  console.log('│ Agregar al INICIO      │      O(1)       │     O(n)      │');
  console.log('│ Agregar al FINAL       │      O(n)       │     O(1)*     │');
  console.log('│ Quitar del INICIO      │      O(1)       │     O(n)      │');
  console.log('│ Quitar del FINAL       │      O(n)       │     O(1)      │');
  console.log('│ Acceso por índice      │      O(n)       │     O(1)      │');
  console.log('│ Compartir datos        │      O(1)       │     O(n)**    │');
  console.log('═'.repeat(65));
  console.log('* Amortizado');
  console.log('** Requiere copia defensiva para seguridad');
  console.log();
}

// =============================================================================
// DEMOSTRACIÓN PRINCIPAL
// =============================================================================

// console.log('═'.repeat(70));
// console.log('🧠 INMUTABILIDAD INTELIGENTE: EL SECRETO DE LA EFICIENCIA EN FP');
// console.log('═'.repeat(70));
// console.log();

// demoPrependRapido();
// demoInitLento();
// demoTail();
// tablaComplejidad();
// demoCopiasDefensivas();

// console.log('═'.repeat(70));
// console.log('📚 RESUMEN DE CONCEPTOS DEL GRÁFICO:');
// console.log('═'.repeat(70));
// console.log(`
// 1. ESTRUCTURAS INMUTABLES
//    → Una vez creadas, nunca cambian
//    → Las operaciones devuelven NUEVAS estructuras

// 2. COMPARTICIÓN ESTRUCTURAL
//    → No copiamos datos, reutilizamos referencias
//    → Eficiente en memoria y tiempo

// 3. EFICIENCIA PRÁCTICA
//    → Prepend (agregar al inicio): O(1) - Rápido
//    → Init (quitar del final): O(n) - Lento
//    → Diseña tus algoritmos para aprovechar las operaciones rápidas

// 4. ADIÓS COPIAS DEFENSIVAS
//    → Los datos inmutables se pueden compartir sin riesgo
//    → No hay modificaciones inesperadas
//    → Código más simple y seguro
// `);
// console.log('═'.repeat(70));

// =============================================================================
// RETO DE REFACTORIZACIÓN
// =============================================================================

// console.log('\n🎯 RETO DE REFACTORIZACIÓN:');
// console.log('─'.repeat(70));
// console.log(`
// Tienes este código IMPERATIVO/MUTABLE. Piensa cómo lo harías funcional:

// ❌ CÓDIGO MALO (mutable):

class Historial {
    private cambios: string[] = [];

    registrar(cambio: string): void {
        this.cambios.push(cambio);  // Mutación
    }

    deshacer(): void {
        this.cambios.pop();  // Mutación
    }

    getCambios(): string[] {
        return this.cambios;  // ¡Expone el estado!
    }
}

const historial = new Historial();
historial.registrar("Cambio 1");
historial.registrar("Cambio 2");
historial.deshacer();
const changes = historial.getCambios();
changes.push("hack 3")
changes.push("hack 4")
changes[0] = "hack 1"
console.log(changes);


class HistorialInmutable {
    constructor(private readonly cambios: ImmutableList<string> = nil) {
    }

    registrar(cambio: string): HistorialInmutable {
        return new HistorialInmutable(prepend(cambio, this.cambios));
    }

    deshacer(): HistorialInmutable {
        return new HistorialInmutable(tail(this.cambios));
    }

    getCambios(): ImmutableList<string> {
        return this.cambios;
    }
}

let historial2 = new HistorialInmutable();
historial2 = historial2.registrar("Cambio 1");
historial2 = historial2.registrar("Cambio 2");
historial2 = historial2.deshacer();
const changesImmutable = historial2.getCambios();
console.log('[changesImmutable]', changesImmutable);
console.log('[changesImmutable]', toArray(changesImmutable));
// changesImmutable.push("hack 3")
// changesImmutable._tag = "Nil"
// changesImmutable.head = "Nil"
// changesImmutable.tail = "Nil"


// 💭 PREGUNTA: 
//    ¿Cómo implementarías este Historial de forma inmutable usando
//    una lista enlazada? ¿Qué operaciones serían O(1)?

//    Pista: "registrar" debería ser prepend, "deshacer" debería ser tail.
// `);
// console.log('─'.repeat(70));

function length(lista: ImmutableList<number>): number {
  if (lista._tag === 'Nil') {
    return 0; // Caso base
  } else {
    // TypeScript ya sabe que aquí 'lista' es 'Cons'
    console.log('[lista]', lista.tail);
    return 1 + length(lista.tail); // Recursión
  }
}

const lista = prepend(3, prepend(2, prepend(1, nil)));
console.log(length(lista)); // 3
console.log(length(nil)); // 0  

export {};
